stdlib/public/core/StringIndex.swift

```diff
@@ -272,6 +301,154 @@ extension String.Index {
+// separate flag is set for UTF-16 indices. (Only foreign strings can be UTF-16
+// encoded. As of 5.7, all foreign strings are UTF-16; but this is subject to
+// change later if we ever decide to implement additional foreign forms.)
+//
+// In releases before 5.7, the bits corresponding to these flags were considered
+// reserved, and they were both set to zero in inlinable code. This means that
+// (on ABI stable platforms at least) we cannot assume that either of these bits
+// will be reliably set. If they are both clear, then we must fall back to
+// assuming that the index has the right encoding for whatever string it is used
+// on. However, if either of these bits are set, then the other bit's value is
+// also reliable -- whether it's set or cleared.
+//
+// The indices of ASCII strings are encoding-independent, i.e. transcoding such
+// strings from UTF-8 to UTF-16 (or vice versa) does not change the position
+// value of any of their indices. Therefore it isn't an error for an index to
+// have both of these flags set. (The start index of every string also behaves
+// this way: position zero is the same no matter how what encoding is used for
+// the rest of string.)
+//
+// These two bits (along with the isForeignUTF8 flag in StringObject) allow
+// newer versions of the Standard Library to more reliably catch runtime errors
+// where client code is applying an index from a UTF-16 string to a UTF-8 one,
+// or vice versa. This typically happens when indices from a UTF-16 Cocoa string
+// that was verbatim bridged into Swift are accidentally applied to a mutated
+// version of the same string. (The mutation turns it into a UTF-8 native
+// string, where the same numerical offsets might correspond to wildly different
+// logical positions.)
+//
+// Such code has always been broken, as the old indices are documented to be no
+// longer valid after the mutation; however, in previous releases such cases
+// weren't reliably detected, and if the code was only ever tested on ASCII
+// strings, then the bug could lie dormant for a long time. (Until the code
+// encounters a non-ASCII character and someone gets surprised that the results
+// no longer make sense.)
+//
+// As more code gets rebuilt with Swift 5.7+, the stdlib will gradually become
+// able to reliably catch and correct all such issues. The error cases are
+// handled in `_StringGuts.ensureMatchingEncoding(_:)`; see there for the sordid
+// details.
+extension String.Index {
+  /// Returns true if the position in this index can be interpreted as an offset
+  /// into UTF-8-encoded string storage.
+  ///
+  /// (This returns true if either we know for sure that this is an UTF-8 index,
+  /// or if we don't have enough information to determine its encoding.)
+  @_alwaysEmitIntoClient // Swift 5.7
+  @inline(__always)
+  internal var _canBeUTF8: Bool {
+    // The only way an index cannot be UTF-8 is it has only the UTF-16 flag set.
+    _rawBits & 0xC != 0x08
```

Comment:
Aside: We need a better solution for bitfield, flag set, what name have you. `OptionSet` is ok, but I think we can do better...

Reply:
I made this a bit more readable by assigning symbolic names to the individual bits. This now looks like this (without all the fluff):

```swift
  internal var _encodingBits: UInt64 { _rawBits & (Self.__utf8Bit | Self.__utf16Bit) }
  internal var _canBeUTF8: Bool { _encodingBits != Self.__utf16Bit }
```